packages/compass-import-export/src/import/import-csv.ts

```diff
@@ -0,0 +1,182 @@
+  let parsedHeader: Record<string, PathPart[]>;
+
+  const docStream = new Transform({
+    objectMode: true,
+    transform: function (chunk: Record<string, string>, encoding, callback) {
+      if (!parsedHeader) {
+        parsedHeader = {};
+        for (const [index, name] of headerFields.entries()) {
+          try {
+            parsedHeader[name] = parseHeaderName(name);
+          } catch (err: unknown) {
+            // rethrow with the row and column indexes appended to aid debugging
+            // TODO: this needs a test
+            (err as Error).message = `${
+              (err as Error).message
+            }[Row 0][Col ${index}]`;
+            debug('parseHeaderName error', (err as Error).message);
+
+            // If this fails, the whole file will stop processing regardless of
+            // the value of stopOnErrors because it is not recoverable if we
+            // can't make sense of the header row.
+            return callback(err as Error);
+          }
+        }
+      }
+
+      // Call progress and increase the number processed even if it errors
+      // below. The collection write stream stats at the end stores how many
+      // got written. This way progress updates continue even if every row
+      // fails to parse.
+      ++numProcessed;
+      progressCallback?.(numProcessed);
+
+      debug('importCSV:transform', numProcessed, {
+        headerFields,
+        chunk,
+        fields,
+        ignoreEmptyStrings,
+        encoding,
+      });
+
+      try {
+        const doc = makeDoc(chunk, headerFields, parsedHeader, fields, {
+          ignoreEmptyStrings,
+        });
+        debug('transform', doc);
+        callback(null, doc);
+      } catch (err: unknown) {
+        // rethrow with the row index appended to aid debugging
+        (err as Error).message = `${
+          (err as Error).message
+        }[Row ${numProcessed}]`;
+
+        if (stopOnErrors) {
+          callback(err as Error);
+        } else {
+          const transformedError = errorToJSON(err);
+          debug('transform error', transformedError);
+          errorCallback?.(transformedError);
+          output.write(JSON.stringify(transformedError) + os.EOL, 'utf8', () =>
+            callback()
+          );
+        }
+      }
+    },
+  });
+
+  const collectionStream = createCollectionWriteStream(
+    dataService,
+    ns,
+    stopOnErrors ?? false
+  );
+
+  const parseStream = Papa.parse(Papa.NODE_STREAM_INPUT, {
+    delimiter,
+    header: true,
+    transformHeader: function (header: string, index: number): string {
+      debug('importCSV:transformHeader', header, index);
+      headerFields.push(header);
+      return header;
+    },
+  });
+
+  const params = [
+    input,
+    parseStream,
+    docStream,
+    collectionStream,
+    ...(abortSignal ? [{ signal: abortSignal }] : []),
+  ] as const;
+
+  // This is temporary until we change WritableCollectionStream so it can pipe
+  // us its errors as they occur.
```

Comment:
☝️ Will tackle this when I fix the existing stopOnErrors bug in the collection write stream. In a separate PR.